src/subprocess-posix.cc

```diff
@@ -169,11 +171,34 @@ void SubprocessSet::HandlePendingInterruption() {
     interrupted_ = SIGTERM;
 }
 
+void SubprocessSet::HandleChildDeath() {
+  if (interrupted_ != SIGCHLD)
+    return;
+  interrupted_ = false;
+  int options = WEXITED | WNOWAIT | WNOHANG;
+  for (vector<Subprocess*>::iterator i = running_.begin();
+       i != running_.end(); ) {
+    siginfo_t info = {0};
+    int res = waitid(P_PID, (id_t)(*i)->pid_, &info, options);
+    if (res == -1)
+      Fatal("waitid(%d): %s", (*i)->pid_, strerror(errno));
+    if (info.si_pid == 0) {
+      ++i;
+    } else {
+      close((*i)->fd_);
```

Comment:
If HandleChildDeath is called before we read all active fds as it is now, doing close() could discard several last lines of process output, I think.